src/Data/List.purs

```diff
@@ -257,6 +258,13 @@ uncons :: forall a. List a -> Maybe { head :: a, tail :: List a }
 uncons Nil = Nothing
 uncons (x : xs) = Just { head: x, tail: xs }
 
+-- | Break a list into its last element, and the preceding elements,
+-- | or `Nothing` if the list is empty.
+-- |
+-- | Running time: `O(2n)`
+unsnoc :: forall a. List a -> Maybe { init :: List a, last :: a }
+unsnoc xs = { init: _, last: _ } <$> init xs <*> last xs
```

Comment:
My motivation for adding it was that it makes pattern-matching cleaner. Otherwise, you have to do separate `Maybe` checks for both `init` and `last`. But sure, I'd be happy to optimize it. 😃